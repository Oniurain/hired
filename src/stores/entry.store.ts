import { observable, computed, action, reaction} from 'mobx';
import * as moment from 'moment';
import * as uuid from 'uuid';

import { Fb, StorageEntryInterface } from '../storage/firebase';

class EntryStore {
  @observable startTime: Date;
  @observable endTime: Date;
  @observable seconds = 0;
  tickInterval: any;
  id: String;

  static fromStorage(id: string, storageObject: StorageEntryInterface) {
    return new this(id, storageObject.startTime, storageObject.endTime);
  }

  constructor(id: string = uuid.v4(), startTime?: number, endTime?: number|null) {
    this.id = id;

    if (startTime) {
      this.startTime = new Date(startTime);
    }

    if (endTime) {
      this.endTime = new Date(endTime);
    }

    reaction(
      () => ([this.startTime, this.endTime]),
      (changes) => {
        // tslint:disable-next-line
        console.log('in entry reaction', changes);
        Fb.database.ref(`entries/${this.id}`).set(this.toStorage());
      },
      {
        fireImmediately: true
      }
    );
  }

  @computed get running() {
    return !this.endTime;
  }

  @computed get duration() {
    let start = moment(this.startTime);

    if (this.running) {
      return this.seconds;
    } else {
      let end = moment(this.endTime);
      return moment.duration(end.diff(start)).asSeconds();
    }
  }

  @action
  startTimer(start: Date = new Date()) {
    this.startTime = start;

    this.tickInterval = setInterval(
      () => {
        this.tick();
      },
      1000
    );
  }

  @action
  stopTimer(end: Date = new Date()) {
    this.endTime = end;
    clearInterval(this.tickInterval);
  }

  toStorage(): StorageEntryInterface {
    return {
      startTime: moment(this.startTime).valueOf(),
      endTime: (this.endTime) ? moment(this.endTime).valueOf() : null
    };
  }

  @action
  private tick() {
    this.seconds += 1;
  }

}

export { EntryStore };
