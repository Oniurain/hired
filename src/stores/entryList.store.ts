import { observable, action } from 'mobx'
import { sortBy } from 'lodash'

import { EntryStore } from './entry.store'

class EntryListStore {
  @observable entries: EntryStore[] = []
  @observable active: EntryStore | null = null

  constructor(entries = {}) {
    if (entries) {
      this.hydrate(entries)
    }
  }

  @action
  addEntry(entry: EntryStore) {
    this.entries.unshift(entry)
  }

  @action
  startNewEntry() {
    if (!this.active) {
      let entry = new EntryStore()
      this.active = entry
      this.addEntry(entry)
      entry.startTimer()
    }
  }

  @action
  stopCurrentTimer() {
    if (this.active) {
      this.active.stopTimer()
      this.active = null
    }
  }

  @action
  private setEntries(entries: any[]) {
    this.entries = entries
  }

  @action
  private checkFirstActive() {
    console.log('in checkFirstActive', this.entries)
    if (this.entries[0].running) {
      this.active = this.entries[0]
    }
  }

  private hydrate(storageObjects: Object) {
    let unsortedEntries = []

    for (let id in storageObjects) {
      unsortedEntries.push(EntryStore.fromStorage(id, storageObjects[id]))
    }

    let sortedEntries = sortBy(unsortedEntries, ['startTime']).reverse()

    this.setEntries(sortedEntries)
    this.checkFirstActive()
  }
}

export { EntryListStore }
