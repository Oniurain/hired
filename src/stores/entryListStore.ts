import { observable, action, computed, reaction } from 'mobx'
import { sortBy } from 'lodash'
import * as uuid from 'uuid'

import { StorageEntryInterface } from '../storage/firebaseRepository'
import RootStore from './rootStore'
import EntryStore from './entryStore'

class EntryListStore {
  @observable entries: EntryStore[] = []
  @observable active: EntryStore | null = null

  rootStore: RootStore

  constructor(rootStore: RootStore) {
    this.rootStore = rootStore

    this.setupSync()
  }

  @action
  addEntry(entry: EntryStore) {
    this.entries.unshift(entry)
  }

  @action.bound
  startNewEntry() {
    if (this.active) {
      this.stopCurrentTimer()
    }

    let entry = new EntryStore({
      id: uuid.v4(),
      projectId: this.rootStore.projectListStore.currentProject.id,
      rate: this.rootStore.projectListStore.currentProject.standardRate
    })
    this.active = entry
    this.addEntry(entry)
    entry.startTimer()
  }

  @action.bound
  stopCurrentTimer() {
    if (this.active) {
      this.active.stopTimer()
      this.active = null
    }
  }

  getEntriesForProject(projectId: string): EntryStore[] {
    return this.entries.filter((entry: EntryStore) => (entry.projectId === projectId))
  }

  getFinishedEntriesForProject(projectId: string): EntryStore[] {
    return this.getEntriesForProject(projectId).filter((entry: EntryStore) => (!entry.running))
  }

  @computed get toStorage() {
    return this.entries.reduce(
      (map, entry) => (
        Object.assign({}, map, { [entry.id]: entry.toStorage })
      ),
      {}
    )
  }

  @action.bound
  hydrate(storageObjects: StorageEntryInterface[]) {
    const unsortedEntries = storageObjects.map((item) => EntryStore.fromStorage(item))
    const sortedEntries = sortBy(unsortedEntries, ['startTime']).reverse()

    this.setEntries(sortedEntries)
    this.checkFirstActive()
  }

  @action.bound
  private setEntries(entries: EntryStore[]) {
    this.entries = entries
  }

  @action.bound
  private checkFirstActive() {
    if (this.entries[0].running) {
      this.active = this.entries[0]
    }
  }

 private setupSync(): void {
    reaction(
      () => this.toStorage,
      (entries) => {
        if (entries) {
          this.rootStore.repository.entries(this.rootStore.userStore.uid).set(entries)
        }
      }
    )
  }
}

export default EntryListStore
