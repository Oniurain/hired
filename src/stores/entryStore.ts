import { observable, computed, action } from 'mobx'
import { now } from 'mobx-utils'
import * as moment from 'moment'
import * as uuid from 'uuid'

import { StorageEntryInterface } from '../storage/firebaseRepository'

const STANDARD_RATE = 20

class EntryStore {
  @observable startTime: Date
  @observable endTime: Date
  @observable seconds = 0
  rate: number
  tickInterval: any
  id: string
  projectId: string

  static fromStorage(id: string, storageObject: StorageEntryInterface) {
    return new this({
      id,
      startTime: storageObject.startTime,
      endTime: storageObject.endTime,
      rate: storageObject.rate,
      projectId: storageObject.projectId
    })
  }

  constructor({
    id = uuid.v4(),
    startTime,
    endTime,
    rate = STANDARD_RATE,
    projectId
  }: { id?: string, startTime?: number, endTime?: number | null, rate: number, projectId: string }) {
    this.id = id

    if (startTime) {
      this.startTime = new Date(startTime)
    }

    if (endTime) {
      this.endTime = new Date(endTime)
    }

    if (startTime && !endTime) {
      this.startTimer(this.startTime)
    }

    this.rate = rate

    this.projectId = projectId
  }

  @computed get running() {
    return !this.endTime
  }

  @computed get duration() {
    let end = (this.endTime) ? this.endTime.getTime() : now()
    return (end - this.startTime.getTime()) / 1000
  }

  @computed get total() {
    return (this.rate / 3600) * this.duration
  }

  @action
  startTimer(start: Date = new Date()) {
    this.startTime = start
  }

  @action
  stopTimer(end: Date = new Date()) {
    this.endTime = end
    clearInterval(this.tickInterval)
  }

  toStorage(): StorageEntryInterface {
    return {
      id: this.id,
      startTime: moment(this.startTime).valueOf(),
      endTime: (this.endTime) ? moment(this.endTime).valueOf() : null,
      rate: this.rate,
      projectId: this.projectId
    }
  }

}

export { EntryStore }
