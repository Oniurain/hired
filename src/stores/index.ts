import EntryListStore from './entryListStore'
import EntryStore from './entryStore'
import ProjectListStore from './projectListStore'
import ProjectStore from './projectStore'
import RootStore from './rootStore'
import UiStore from './uiStore'
import UserStore from './userStore'

export {
  EntryListStore,
  EntryStore,
  ProjectListStore,
  ProjectStore,
  RootStore,
  UiStore,
  UserStore
}
