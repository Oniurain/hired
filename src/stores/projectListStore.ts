import { observable, action, computed, reaction } from 'mobx'
import * as uuid from 'uuid'

import RootStore from './rootStore'
import ProjectStore, { ProjectInterface } from './projectStore'

class ProjectListStore {
  @observable projects: ProjectStore[] = []

  private rootStore: RootStore

  constructor(rootStore: RootStore) {
    this.rootStore = rootStore

    this.setupSync()
  }

  @action
  hydrate(projects: ProjectInterface[]) {
    this.projects = projects.map((project) => {
      return new ProjectStore(this.rootStore, project.id, project.name, project.standardRate, project.description)
    })
  }

  @action
  add(name: string, standardRate: number, description: string = '', id: string = uuid.v4()) {
    this.projects = [...this.projects, new ProjectStore(this.rootStore, id, name, standardRate, description)]
  }

  @computed get currentProject() {
    return this.projects.filter((item) => (item.id === this.rootStore.uiStore.currentView.projectId))[0]
  }

  getById(id: string): ProjectStore {
    return this.projects.filter((item) => (item.id === id))[0]
  }

  private setupSync() {
    reaction(
      () => this.projects.map((project) => project.toStorage),
      (projects) => {
        if (projects) {
          this.rootStore.repository.projects(this.rootStore.userStore.uid).set(projects)
        }
      }
    )
  }

}

export default ProjectListStore
