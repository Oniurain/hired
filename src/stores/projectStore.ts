import { observable, action } from 'mobx'

class ProjectStore {
  @observable isFetched: boolean = false
  @observable projects: Object = {}
  @observable currentProject: ProjectInterface

  // @action
  // fetchProjects() {

  // }

  @action
  add(id: string, name: string, rate: number, description: string = '') {
    const project = {
      id,
      name,
      rate,
      description
    }

    this.projects[id] = project
  }

  @action
  setCurrent(id: string) {
    if (this.projects[id]) {
      this.currentProject = this.projects[id]
    } else {
      throw new Error(`Project with ${id} does not exist`)
    }
  }

}

export interface ProjectInterface {
  id: string,
  name: string,
  rate: number,
  description: string
}

const projectStore = new ProjectStore()

// TODO: dummy
projectStore.add('neusta', 'Neusta', 67.50, 'A super cool project')
projectStore.setCurrent('neusta')


export { ProjectStore }
export default projectStore
