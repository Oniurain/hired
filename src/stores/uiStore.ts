import { observable, action } from 'mobx'

interface CurrentViewInterface {
  name: string,
  projectId?: string
}

const VIEW_NAMES = {
  projectList: 'PROJECT_LIST',
  projectOverview: 'PROJECT_OVERVIEW',
  projectNew: 'PROJECT_NEW'
}

class UiStore {
  @observable isLoaded: boolean = false
  @observable hasError: boolean = false
  @observable currentView: CurrentViewInterface = {
    name: VIEW_NAMES.projectList
  }

  @action
  showProjectList(): void {
    this.currentView = {
      name: VIEW_NAMES.projectList
    }
  }

  @action
  showProject(projectId: string): void {
    this.currentView = {
      name: VIEW_NAMES.projectOverview,
      projectId
    }
  }

  @action
  showNewProject(): void {
    this.currentView = {
      name: VIEW_NAMES.projectNew
    }
  }

  @action
  setLoaded (isLoaded: boolean): void {
    this.isLoaded = isLoaded
  }

  @action
  setError (status: boolean): void {
    this.hasError = status
  }
}

export default UiStore
export { VIEW_NAMES }
