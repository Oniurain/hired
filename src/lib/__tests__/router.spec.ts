import { createRootStore } from '../../../test/util'
import { RootStore } from '../../stores'
import Router from '../router'
import { VIEW_NAMES } from '../../stores/uiStore'

import Router5 from 'router5'

describe('Router', () => {
  let rootStore: RootStore
  let subject: Router

  beforeEach(() => {
    rootStore = createRootStore()

    subject = new Router(rootStore)
  })

  it('should initialize', () => {
    expect(subject.instance).toMatchSnapshot()
  })

  it('should navigate to to new project screen', () => {
    subject.navigate(VIEW_NAMES.projectNew)

    expect(rootStore.uiStore.hasError).toBe(false)
    expect(rootStore.uiStore.currentView.name).toBe(VIEW_NAMES.projectNew)
  })

  it('should navigate to to project-list-screen', () => {
    subject.navigate(VIEW_NAMES.projectList)

    expect(rootStore.uiStore.hasError).toBe(false)
    expect(rootStore.uiStore.currentView.name).toBe(VIEW_NAMES.projectList)
  })

  it('should navigate to to project-overview-screen', () => {
    subject.navigate(VIEW_NAMES.projectOverview, { projectId: '1234-5678-123'})

    expect(rootStore.uiStore.hasError).toBe(false)
    expect(rootStore.uiStore.currentView.name).toBe(VIEW_NAMES.projectOverview)
    expect(rootStore.uiStore.currentView.projectId).toBe('1234-5678-123')
  })

  it('should enter error state when navigation to wrong route was attempted', () => {
    subject.navigate('foo')

    expect(rootStore.uiStore.hasError).toBe(true)
  })
})
