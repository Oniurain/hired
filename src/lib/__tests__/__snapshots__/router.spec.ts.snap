// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Router should initialize 1`] = `
Object {
  "add": [Function],
  "addEventListener": [Function],
  "addListener": [Function],
  "addNode": [Function],
  "addNodeListener": [Function],
  "addRouteListener": [Function],
  "areStatesDescendants": [Function],
  "areStatesEqual": [Function],
  "buildPath": [Function],
  "buildState": [Function],
  "buildUrl": [Function],
  "canActivate": [Function],
  "canDeactivate": [Function],
  "cancel": [Function],
  "clearCanDeactivate": [Function],
  "clearMiddleware": [Function],
  "clone": [Function],
  "executeFactory": [Function],
  "getDependencies": [Function],
  "getLifecycleFactories": [Function],
  "getLifecycleFunctions": [Function],
  "getListeners": [Function],
  "getMiddlewareFactories": [Function],
  "getMiddlewareFunctions": [Function],
  "getOptions": [Function],
  "getPlugins": [Function],
  "getState": [Function],
  "hasPlugin": [Function],
  "invokeEventListeners": [Function],
  "isActive": [Function],
  "isStarted": [Function],
  "makeNotFoundState": [Function],
  "makeState": [Function],
  "matchPath": [Function],
  "matchUrl": [Function],
  "navigate": [Function],
  "navigateToDefault": [Function],
  "removeEventListener": [Function],
  "removeListener": [Function],
  "removeNodeListener": [Function],
  "removeRouteListener": [Function],
  "replaceHistoryState": [Function],
  "rootNode": RouteNode {
    "absolute": false,
    "children": Array [
      RouteNode {
        "absolute": false,
        "children": Array [],
        "name": "PROJECT_NEW",
        "parent": [Circular],
        "parser": Path {
          "hasMatrixParams": false,
          "hasQueryParams": false,
          "hasSpatParam": false,
          "hasUrlParams": false,
          "params": Array [],
          "path": "/project/new",
          "queryParams": Array [],
          "queryParamsBr": Array [],
          "source": "\\\\/project\\\\/new",
          "spatParams": Array [],
          "tokens": Array [
            Object {
              "match": "/",
              "otherVal": Array [],
              "regex": /\\\\//,
              "type": "delimiter",
              "val": Array [
                "/",
              ],
            },
            Object {
              "match": "project",
              "otherVal": Array [],
              "regex": /project/,
              "type": "fragment",
              "val": Array [
                "project",
              ],
            },
            Object {
              "match": "/",
              "otherVal": Array [],
              "regex": /\\\\//,
              "type": "delimiter",
              "val": Array [
                "/",
              ],
            },
            Object {
              "match": "new",
              "otherVal": Array [],
              "regex": /new/,
              "type": "fragment",
              "val": Array [
                "new",
              ],
            },
          ],
          "urlParams": Array [],
        },
        "path": "/project/new",
      },
      RouteNode {
        "absolute": false,
        "children": Array [],
        "name": "PROJECT_OVERVIEW",
        "parent": [Circular],
        "parser": Path {
          "hasMatrixParams": false,
          "hasQueryParams": false,
          "hasSpatParam": false,
          "hasUrlParams": true,
          "params": Array [
            "projectId",
          ],
          "path": "/project/:projectId",
          "queryParams": Array [],
          "queryParamsBr": Array [],
          "source": "\\\\/project\\\\/([a-zA-Z0-9-_.~%':]+)",
          "spatParams": Array [],
          "tokens": Array [
            Object {
              "match": "/",
              "otherVal": Array [],
              "regex": /\\\\//,
              "type": "delimiter",
              "val": Array [
                "/",
              ],
            },
            Object {
              "match": "project",
              "otherVal": Array [],
              "regex": /project/,
              "type": "fragment",
              "val": Array [
                "project",
              ],
            },
            Object {
              "match": "/",
              "otherVal": Array [],
              "regex": /\\\\//,
              "type": "delimiter",
              "val": Array [
                "/",
              ],
            },
            Object {
              "match": ":projectId",
              "otherVal": Array [
                undefined,
                undefined,
              ],
              "regex": /\\(\\[a-zA-Z0-9-_\\.~%':\\]\\+\\)/,
              "type": "url-parameter",
              "val": Array [
                "projectId",
              ],
            },
          ],
          "urlParams": Array [
            "projectId",
          ],
        },
        "path": "/project/:projectId",
      },
      RouteNode {
        "absolute": false,
        "children": Array [],
        "name": "PROJECT_LIST",
        "parent": [Circular],
        "parser": Path {
          "hasMatrixParams": false,
          "hasQueryParams": false,
          "hasSpatParam": false,
          "hasUrlParams": false,
          "params": Array [],
          "path": "/",
          "queryParams": Array [],
          "queryParamsBr": Array [],
          "source": "\\\\/",
          "spatParams": Array [],
          "tokens": Array [
            Object {
              "match": "/",
              "otherVal": Array [],
              "regex": /\\\\//,
              "type": "delimiter",
              "val": Array [
                "/",
              ],
            },
          ],
          "urlParams": Array [],
        },
        "path": "/",
      },
    ],
    "name": "",
    "parent": undefined,
    "parser": null,
    "path": "",
  },
  "setDependencies": [Function],
  "setDependency": [Function],
  "setOption": [Function],
  "setRootPath": [Function],
  "setState": [Function],
  "start": [Function],
  "stop": [Function],
  "transitionToState": [Function],
  "urlToPath": [Function],
  "useMiddleware": [Function],
  "usePlugin": [Function],
}
`;
